src/vsr/journal.zig

```diff
@@ -2142,25 +2145,34 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
             return sector;
         }
 
-        pub fn writing(journal: *Journal, op: u64, checksum: u128) bool {
-            const slot = journal.slot_for_op(op);
-            var found: bool = false;
-            var it = journal.writes.iterate();
-            while (it.next()) |write| {
-                const write_slot = journal.slot_for_op(write.message.header.op);
+        const Writing = enum {
+            none,
+            /// Either the prepare or the redundant header of a message with the same slot as the
+            /// given op is being written. It may be a different version of the same op, or a
+            /// different op which shares the prepare slot.
+            slot,
+            /// Either the prepare or the redundant header of a message with the exact op/checksum
+            /// is being written.
+            exact,
+        };
 
-                // It's possible that we might be writing the same op but with a different checksum.
-                // For example, if the op we are writing did not survive the view change and was
-                // replaced by another op. We must therefore do the search primarily on checksum.
-                // However, we compare against the 64-bit op first, since it's a cheap machine word.
-                if (write.message.header.op == op and write.message.header.checksum == checksum) {
-                    // If we truly are writing, then the dirty bit must be set:
-                    assert(journal.dirty.bit(journal.slot_for_op(op)));
-                    found = true;
-                } else if (write_slot.index == slot.index) {
-                    // If the in-progress write of '{op, checksum}' will be overwritten by another
-                    // write to the same slot, writing() must return false.
-                    found = false;
+        pub fn writing(journal: *Journal, header: *const Header.Prepare) Writing {
+            const slot = journal.slot_for_header(header);
+            var found: Writing = .none;
+            var writes = journal.writes.iterate();
+            while (writes.next()) |write| {
+                const write_slot = journal.slot_for_op(write.message.header.op);
+                if (write.message.header.op == header.op and
+                    write.message.header.checksum == header.checksum)
+                {
+                    assert(write_slot.index == slot.index);
+                    assert(found == .none);
+                    found = .exact;
+                } else {
+                    if (write_slot.index == slot.index) {
+                        assert(found == .none);
+                        found = .slot;
+                    }
                 }
```

Comment:
Took me a moment to convince myself that the nested ifs don't miss a different prepare in the same slot. Perhaps we can flatten this?

```zig
                if (write_slot.index == slot.index) {
                    assert(found == .none);
                    found = .slot;
                }
                if (write.message.header.op == header.op and
                    write.message.header.checksum == header.checksum)
                {
                    assert(write_slot.index == slot.index);
                    assert(found == .slot);
                    found = .exact;
                }
```

Not entirely sure if that's better though!

Reply:
Replaced with:

```zig
                if (write_slot.index == slot.index) {
                    assert(found == .none);

                    if (write.message.header.checksum == header.checksum) {
                        assert(write.message.header.op == header.op);
                        found = .exact;
                    } else {
                        maybe(write.message.header.op == header.op);
                        found = .slot;
                    }
                } else {
                    assert(write.message.header.op != header.op);
                }
```

which hopefully makes it more clear!